ogr/ogr_schema_override.h

```diff
@@ -0,0 +1,382 @@
+    {
+        return m_eType;
+    }
+
+    std::optional<OGRFieldSubType> GetFieldSubType() const
+    {
+        return m_eSubType;
+    }
+
+    std::optional<int> GetFieldWidth() const
+    {
+        return m_nWidth;
+    }
+
+    std::optional<int> GetFieldPrecision() const
+    {
+        return m_nPrecision;
+    }
+
+    // Considered valid if it carries any change information, otherwise it's considered a no-op
+    bool IsValid() const
+    {
+        return m_osName.has_value() || m_eType.has_value() ||
+               m_eSubType.has_value() || m_nWidth.has_value() ||
+               m_nPrecision.has_value();
+    }
+
+  private:
+    std::optional<std::string> m_osName;
+    std::optional<OGRFieldType> m_eType;
+    std::optional<OGRFieldSubType> m_eSubType;
+    std::optional<int> m_nWidth;
+    std::optional<int> m_nPrecision;
+};
+
+/** Class that holds the schema override options for a single layer */
+class OGRLayerSchemaOverride
+{
+  public:
+    OGRLayerSchemaOverride() : m_osLayerName(), m_moFieldOverrides()
+    {
+    }
+
+    void SetLayerName(const std::string &osLayerName)
+    {
+        m_osLayerName = osLayerName;
+    }
+
+    void AddFieldOverride(const std::string &osFieldName,
+                          const OGRFieldDefnOverride &oFieldOverride)
+    {
+        m_moFieldOverrides[osFieldName] = oFieldOverride;
+    }
+
+    const std::string &GetLayerName() const
+    {
+        return m_osLayerName;
+    }
+
+    const std::map<std::string, OGRFieldDefnOverride> &GetFieldOverrides() const
+    {
+        return m_moFieldOverrides;
+    }
+
+    bool IsFullOverride() const
+    {
+        return m_bIsFullOverride;
+    }
+
+    void SetFullOverride(bool bIsFullOverride)
+    {
+        m_bIsFullOverride = bIsFullOverride;
+    }
+
+    bool IsValid() const
+    {
+        bool isValid = !m_osLayerName.empty() && !m_moFieldOverrides.empty();
+        for (const auto &oFieldOverride : m_moFieldOverrides)
+        {
+            isValid &= !oFieldOverride.first.empty();
+            // When schemaType is "full" override we don't need to check if the field
+            // overrides are valid: a list of fields to keep is enough.
+            if (!m_bIsFullOverride)
+            {
+                isValid &= oFieldOverride.second.IsValid();
+            }
+        }
+        return isValid;
+    }
+
+  private:
+    std::string m_osLayerName;
+    std::map<std::string, OGRFieldDefnOverride> m_moFieldOverrides;
+    bool m_bIsFullOverride = false;
+};
+
+/** Class that holds the schema override options for a datasource */
+class OGRSchemaOverride
+{
+  public:
+    OGRSchemaOverride() : m_moLayerOverrides()
+    {
+    }
+
+    void AddLayerOverride(const std::string &osLayerName,
+                          const OGRLayerSchemaOverride &oLayerOverride)
+    {
+        m_moLayerOverrides[osLayerName] = oLayerOverride;
+    }
+
+    bool LoadFromJSON(const std::string &osJSON)
```

Comment:
This is a rather long function for a header. Given that this header will be included in multiple compilation units in different drivers, I'd suggest creating a ogr_schema_override.cpp file with the definition of the non-trivial methods of this file

Reply:
done